In [10]:
class EmergencyQueue:
    def __init__(self):
        # Stack to keep track of the most recent patient arrivals
        self.newest_requests = []
        
        # Stack to manage patients in the order they should be served
        self.oldest_requests = []

    def add_request(self, patient_id):
        # A new patient arrives and is added to the waiting stack
        self.newest_requests.append(patient_id)

    def serve_next_patient(self):
        # Make sure the oldest patients are ready to be served
        self._prepare_oldest_requests()

        # Serve (remove) the patient who has been waiting the longest
        if self.oldest_requests:
            self.oldest_requests.pop()

    def peek_next_patient(self):
        # Prepare the stack if it's empty so we can see the next patient
        self._prepare_oldest_requests()

        # Print the ID of the patient who's next in line (without removing them)
        if self.oldest_requests:
            print(self.oldest_requests[-1])

    def _prepare_oldest_requests(self):
        # If there are no patients currently ready to be served,
        # transfer all waiting patients (newest to oldest) into service stack
        if not self.oldest_requests:
            while self.newest_requests:
                self.oldest_requests.append(self.newest_requests.pop())

# Read the total number of instructions (actions)
total_instructions = int(input())

# Create a new hospital emergency queue system
hospital_queue = EmergencyQueue()

# Process each instruction one by one
for i in range(total_instructions):
    instruction = input().strip().split()
    
    if instruction[0] == '1':
        # Instruction type 1: A new patient with a given ID has arrived
        patient_id = int(instruction[1])
        hospital_queue.add_request(patient_id)

    elif instruction[0] == '2':
        # Instruction type 2: The patient at the front of the queue is served
        hospital_queue.serve_next_patient()

    elif instruction[0] == '3':
        # Instruction type 3: We want to see who is next to be treated
        hospital_queue.peek_next_patient()


 8
 1 101
 1 102
 3


101


 2
 3


102


 1 103
 2
 3


103


What’s Happening in the Queue

A new patient with ID 101 arrives and is added to the waiting list.

Soon after, another patient with ID 102 also joins the line.

The doctor takes a quick look to see who's next in line it's patient 101, since they came in first. So, we print 101.

Patient 101 is then called in and treated, so they’re removed from the queue.

The staff checks again to see who's next. Now, it’s patient 102. We print 102.

Another patient, ID 103, arrives and joins the line behind patient 102.

The doctor treats patient 102 next, so they’re removed from the queue.

A final check shows that patient 103 is now first in line. We print 103.

Final output:

101

102

103
